### PR TITLE
Feature/client request properties

### PR DESCRIPTION
**Pull request by danield137**
Exposing ClientRequestProperties
* new class `ClientRequestProperties` that will contain consts / utility methods as time goes on, currently exposes timeout
* added timeout functionality on HttpRequest
* added tests + docs

### CODE REVIEW

File: data/pom.xml [hunk 1]
```diff
@@ -98,7 +98,12 @@
             <version>${mockito.version}</version>
             <scope>test</scope>
         </dependency>
-
+        <dependency>
```
Comment (tamirkamara) [inline data/pom.xml]: Can't find where you're using this...

File: data/src/test/java/com/microsoft/azure/kusto/data/ClientRequestPropertiesTest.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.skyscreamer.jsonassert.JSONAssert;
+
+import java.util.concurrent.TimeUnit;
+
+public class ClientRequestPropertiesTest {
+    @Test
+    @DisplayName("test set/get timeout")
+    void testSetGetTimeout() {
```
Comment (tamirkamara) [inline data/src/test/java/com/microsoft/azure/kusto/data/ClientRequestPropertiesTest.java]: I'd avoid using the test prefix - everything in this class are tests, and also have the "@test" notation...
maybe name it timeout_SetGet

File: data/src/test/java/com/microsoft/azure/kusto/data/ClientRequestPropertiesTest.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        // before setting value should be null
+        Assertions.assertEquals(null, props.getTimeout());
+
+        props.setTimeout(expected);
+        Assertions.assertEquals(expected, props.getTimeout());
+    }
+
+    @Test
+    @DisplayName("test set/get timeout")
+    void testToString() throws JSONException {
```
Comment (tamirkamara) [inline data/src/test/java/com/microsoft/azure/kusto/data/ClientRequestPropertiesTest.java]: I'd avoid using the test prefix - everything in this class are tests, and also have the "@test" notation...
maybe name it timeout_ToString

File: data/src/main/java/com/microsoft/azure/kusto/data/ClientImpl.java [hunk 4]
```diff
@@ -27,31 +27,50 @@ public Results execute(String command) throws DataServiceException, DataClientEx
...
+
+        if (properties != null) {
+            timeoutMs = properties.getTimeout();
+        }
+
         String clusterEndpoint;
         if (command.startsWith(ADMIN_COMMANDS_PREFIX)) {
             clusterEndpoint = String.format("%s/%s/rest/mgmt", clusterUrl, API_VERSION);
+            if (timeoutMs == null) {
+                timeoutMs = TimeUnit.HOURS.toMillis(1) + TimeUnit.SECONDS.toMillis(30);
```
Comment (tamirkamara) [inline data/src/main/java/com/microsoft/azure/kusto/data/ClientImpl.java]: Maybe use the time from a const?

File: data/src/main/java/com/microsoft/azure/kusto/data/ClientImpl.java [hunk 5]
```diff
@@ -27,31 +27,50 @@ public Results execute(String command) throws DataServiceException, DataClientEx
...
-        }
-        return execute(database, command, clusterEndpoint);
-    }
-
-    private Results execute(String database, String command, String clusterEndpoint) throws DataServiceException, DataClientException {
-        // Argument validation:
-        if (StringUtils.isAnyEmpty(database, command, clusterEndpoint)) {
-            throw new IllegalArgumentException("database, command or clusterEndpoint are empty");
+            if (timeoutMs == null) {
+                timeoutMs = TimeUnit.MINUTES.toMillis(4) + TimeUnit.SECONDS.toMillis(30);
```
Comment (tamirkamara) [inline data/src/main/java/com/microsoft/azure/kusto/data/ClientImpl.java]: Maybe use the time from a const?

File: data/src/main/java/com/microsoft/azure/kusto/data/Utils.java [hunk 6]
```diff
@@ -22,8 +24,16 @@
 
 class Utils {
 
-    static Results post(String url, String aadAccessToken, String payload) throws DataServiceException, DataClientException {
-        HttpClient httpClient = HttpClients.createSystem();
+    static Results post(String url, String aadAccessToken, String payload, Long timeoutMs) throws DataServiceException, DataClientException {
+
+        HttpClient httpClient;
+        if (timeoutMs != null) {
+            RequestConfig requestConfig = RequestConfig.custom().setSocketTimeout(timeoutMs.intValue()).build();
```
Comment (tamirkamara) [inline data/src/main/java/com/microsoft/azure/kusto/data/Utils.java]: I get that the socket timeout doesn't allow longs, so maybe we just use int from the start?

File: data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+    public void removeOption(String name) {
+        ((HashMap<String, Object>)properties.get(OPTIONS_KEY)).remove(name);
+    }
+
+    public void clearOptions() {
+        ((HashMap<String, Object>)properties.get(OPTIONS_KEY)).clear();
+    }
+
+    public void setTimeout(Long timeoutInMs) {
```
Comment (tamirkamara) [inline data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java]: I think that we should be explicit here to help the user understand the units - name it setTimeoutInMilliSecs.

File: data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        ((HashMap<String, Object>)properties.get(OPTIONS_KEY)).put(OptionServerTimeout, timeoutInMs);
+    }
+
+    public Long getTimeout() {
```
Comment (tamirkamara) [inline data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java]: rename according to the setter

File: data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+ * Kusto supports attaching various properties to client requests (such as queries and control commands).
+ * Such properties may be used to provide additional information to Kusto (for example, for the purpose of correlating client/service interaction),
+ * may affect what limits and policies get applied to the request, and much more.
+ * For a complete list of available client request properties
+ * check out https://docs.microsoft.com/en-us/azure/kusto/api/netfx/request-properties#list-of-clientrequestproperties
+ */
+public class ClientRequestProperties {
+    private HashMap<String, Object> properties;
+    private static final String OPTIONS_KEY = "Options";
+    private static final String OptionServerTimeout  = "servertimeout";
```
Comment (tamirkamara) [inline data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java]: I think this var should also be all capitals, no?

File: data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.HashMap;
+
+/*
+ * Kusto supports attaching various properties to client requests (such as queries and control commands).
+ * Such properties may be used to provide additional information to Kusto (for example, for the purpose of correlating client/service interaction),
+ * may affect what limits and policies get applied to the request, and much more.
+ * For a complete list of available client request properties
+ * check out https://docs.microsoft.com/en-us/azure/kusto/api/netfx/request-properties#list-of-clientrequestproperties
+ */
+public class ClientRequestProperties {
```
Comment (tamirkamara) [inline data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java]: This class generates many access warnings on the public methods, do we need all of them?
for example - the toJson is used internally, isn't it?

Comment (tamirkamara) [inline data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java]: There're unchecked conversions from Object to HashMap. I know we have that elsewhere, but this is a new class and probably there's a better way to deal with it...
One thing I don't understand about the usage of the HashMap - you have specific key strings defined as finals (inside the options key), meaning that you can do something similar with a well-defined class for the options. can't you?
Something like:
Class ClientRequestPropertiesOptions {
 private int serverTimeoutInMilliSecs;
 private int someOtherOptionKey;
 getters/setters
}

File: data/pom.xml (hunk 1 above)
Comment (danield137) [reply data/pom.xml]: https://github.com/Azure/azure-kusto-java/pull/57/files#diff-dfdcac39245e8c607ad8d656e094db82R32

File: data/src/main/java/com/microsoft/azure/kusto/data/Utils.java (hunk 6 above)
Comment (danield137) [reply data/src/main/java/com/microsoft/azure/kusto/data/Utils.java]: problem is TimeUnit is long, so I either do the conversion here, or on the caller. 
would you prefer it in the caller?

File: data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java (hunk 10 above)
Comment (danield137) [reply data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java]: not sure I understand what you mean..
I only explicitly provided 1 valid option, but there are others many other options.
I want to allow a somewhat dynamic way to manipulate those options, without making code cumbersome.

If you feel there is a better way to do so, please provide a code sample as I couldn't quite get what you meant above 👼

Comment (danield137) [reply data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java]: public is not bad 😄  if you feel any of them should be private, I will change, but client might want to use any of those

**Comment by danield137** [comment]
@tamirkamara fixed most of the comments.
please go over it again when you can 🙏

File: data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java (hunk 10 above)
Comment (tamirkamara) [reply data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java]: Why would I (as a developer) would need it in a json format?

Comment (tamirkamara) [reply data/src/main/java/com/microsoft/azure/kusto/data/ClientRequestProperties.java]: I missed that there're methods that deal with an option with its name as an input.
I suggest to simplify the handling of the hashmaps with something like this:
`
public class ClientRequestProperties {
    private static final String OPTIONS_KEY = "Options";
    private static final String OPTION_SERVER_TIMEOUT = "servertimeout";

    private HashMap<String, Object> properties;
    private HashMap<String, Object> options;

    public ClientRequestProperties() {
        properties = new HashMap<>();
        options = new HashMap<>();
        properties.put(OPTIONS_KEY, options);
    }

    public void setOption(String name, Object value) {
        options.put(name, value);
    }

    public Object getOption(String name) {
        return options.get(name);
    }

    public void removeOption(String name) {
        options.remove(name);
    }

    public void clearOptions() {
        options.clear();
    }

    public Long getTimeoutInMilliSec() {
        return (Long) getOption(OPTION_SERVER_TIMEOUT);
    }

    public void setTimeoutInMilliSec(Long timeoutInMs) {
        setOption(OPTION_SERVER_TIMEOUT, timeoutInMs);
    }

    public JSONObject toJson() {
        return new JSONObject(properties);
    }

    public String toString() {
        return toJson().toString();
    }
}
`

File: data/src/test/java/com/microsoft/azure/kusto/data/ClientRequestPropertiesTest.java (hunk 3 above)
Comment (tamirkamara) [reply data/src/test/java/com/microsoft/azure/kusto/data/ClientRequestPropertiesTest.java]: The link is explaining display names, not overall naming convention. I remember us checking this in the past and reached the conclusion that it is redundant. In this case for JUnit it isn't required in the method names...
https://dzone.com/articles/7-popular-unit-test-naming
https://www.petrikainulainen.net/programming/testing/writing-clean-tests-naming-matters/

I don't really mind one way or the other, but we didn't do this until now.

**Comment by tamirkamara** [comment]
also, this PR should addresses #40, right?
If yes, maybe include that comment in the PR description...

**Pull request merged**